src/mococrw/distinguished_name.h

```diff
@@ -80,19 +82,17 @@ class DistinguishedName
 
     static DistinguishedName fromX509Name(X509_NAME *ptr);
 
+    const std::string& getSubjectAttributeByNIDAsString(const openssl::ASN1_NID id) const;
+    const std::deque<SubjectAttribute>& getSubject() const { return _subjectAttributes; }
+
     bool operator==(const DistinguishedName& other) const;
     bool operator!=(const DistinguishedName& other) const { return !(*this == other); }
-private:
-    std::string _commonName;
-    std::string _countryName;
-    std::string _localityName;
-    std::string _stateOrProvinceName;
-    std::string _organizationalUnitName;
-    std::string _organizationName;
-    std::string _pkcs9EmailAddress;
-    std::string _serialNumber;
-    std::string _givenName;
-    std::string _userId;
+private:    
+    const SubjectAttribute& _getSubjectAttributeByNID(const openssl::ASN1_NID id) const;
+    void _addString(openssl::SSL_X509_NAME_Ptr &x509Name, const SubjectAttribute& unit) const;
+    std::deque<SubjectAttribute> _subjectAttributes;
+    SubjectAttribute _emptyAttribute;
```

Comment:
So that we can use const reference and dont have to return a copy of std::string every time.

Reply:
Sorry but this is not good design. Default objects are kind of an anti-pattern and can lead to a multitude of issues. Moving strings around and having empty strings is expensive enough so that you don't need to get into this kind of design nightmares.
Moreover, there is return-value-optimization.  So just return values instead of const refs. What is more, if the attribute is empty, which means that it doesn't exist, you can just return an optional<> none.

Reply:
DONE